Neos.ContentRepository.Core/Classes/Projection/CatchUpHookFactoryInterface.php

```diff
@@ -4,12 +4,21 @@
 
 namespace Neos\ContentRepository\Core\Projection;
 
-use Neos\ContentRepository\Core\ContentRepository;
+use Neos\ContentRepository\Core\SharedModel\ContentRepository\ContentRepositoryId;
 
 /**
+ * @template T of ProjectionStateInterface
  * @api
  */
 interface CatchUpHookFactoryInterface
 {
-    public function build(ContentRepository $contentRepository): CatchUpHookInterface;
+    /**
+     * Note that a catchup doesn't have access to the full content repository, as it would allow full recursion via handle and accessing other projections
+     * state is not safe as the other projection might not be behind - the order is undefined.
+     *
+     * @param ContentRepositoryId $contentRepositoryId the content repository the catchup was registered in
+     * @param ProjectionStateInterface&T $projectionState the state of the projection the catchup was registered to (Its only safe to access this projections state)
+     * @return CatchUpHookInterface
+     */
+    public function build(ContentRepositoryId $contentRepositoryId, ProjectionStateInterface $projectionState): CatchUpHookInterface;
```

Comment:
maybe we should rather go the way of introducing `CatchUpHookFactoryDependencies` so we can flexibly add new stuff like the `NodeTypeManger` and maybe the content dimension source and other stuff :)

Also as a catchup is build every time new for every catchup of events i think we can pass the the state of if we are booding or just catchuping normally here into the `build` function as well?

As alternative it would need to be handled via ->onBefore and stored internally that way